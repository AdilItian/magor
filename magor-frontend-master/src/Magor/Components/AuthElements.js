import React, { useState, useEffect } from 'react'
import { useHistory } from 'react-router-dom'

import { DropDown } from './Inputs'

import dataProvider from '../../dataProvider'
import authProvider from '../../authProvider'

import { UploadModal } from './UploadModal'

export const paths = {
    Home: '/',
    'Upload Panel': '/upload',
    'My Uploads': '/recordingsManager',
    'Admin Panel': '/admin/',
    'Log Out': '/login',
}

export const UserDropdown = (props) => {
    const { currentPath } = props

    const [user, setUser] = useState(null)
    const [show, setShow] = useState(false)
    const handleClose = () => setShow(false)
    const handleShow = () => setShow(true)

    const history = useHistory()

    useEffect(() => {
        dataProvider
            .getCustom('')
            .then((res) => {
                res.name && setUser(res)
            })
            .catch((_) => history.push('/login'))
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [])

    const logout = (_) => {
        authProvider.logout().then((_) => history.push('/login'))
    }

    if (!user || user.error) return null

    const { name, role } = user
    const admin = role === 'admin'
    const uploader = admin || role === 'uploader'
    const list = []
    list.push('Home')
    if (admin) list.push('Upload Media')
    // if (uploader && currentPath !== 'upload') list.push('Upload Panel')
    if (uploader && currentPath !== 'recordingsManager') list.push('My Uploads')
    if (admin) list.push('Admin Panel')
    list.push('Log Out')

    const performAction = (a) => {
        switch (a) {
            case 'Log Out':
                logout()
                break
            case 'Upload Media':
                handleShow()
                break
            default:
                history.push(paths[a])
        }
    }

    return (
        <>
            <DropDown list={list} selected={name} setSelected={performAction} />
            <UploadModal
                show={show}
                handleShow={handleShow}
                handleClose={handleClose}
            />
        </>
    )
}
