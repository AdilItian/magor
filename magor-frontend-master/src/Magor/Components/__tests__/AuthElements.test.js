/* eslint-disable import/first */
jest.mock('../../../dataProvider')
import React from 'react'
import {MemoryRouter, Router} from 'react-router-dom'
import {createMemoryHistory} from 'history'
import {UserDropdown, paths} from '../AuthElements'
import renderer from 'react-test-renderer'

beforeEach(() => {
    localStorage.clear();
})

test('No undefined', () => {
    expect(UserDropdown).toBeDefined()
    expect(paths).toBeDefined()
})

test('Basic tests', async () => {
    let menu;
    await renderer.act(async () => {
        menu = renderer.create(
            <MemoryRouter>
                <UserDropdown />
            </MemoryRouter>
        )
    })
    expect(menu.toJSON()).toMatchSnapshot()
})

test('Test different auth types', async () => {
    let menu;
    localStorage.setItem('role', 'admin')
    await renderer.act(async () => {
        menu = renderer.create(
            <MemoryRouter>
                <UserDropdown />
            </MemoryRouter>
        )
    })
    expect(menu.root.findAllByType('li').map((li) => li.children[0])).toEqual([
        'Home',
        'Upload Panel',
        'My Uploads',
        'Admin Panel',
        'Log Out',
    ])

    localStorage.setItem('role', 'uploader')
    await renderer.act(async () => {
        menu = renderer.create(
            <MemoryRouter>
                <UserDropdown />
            </MemoryRouter>
        )
    })
    expect(menu.root.findAllByType('li').map((li) => li.children[0])).toEqual([
        'Home',
        'Upload Panel',
        'My Uploads',
        // 'Admin Panel',
        'Log Out',
    ])

    localStorage.setItem('role', 'user')
    await renderer.act(async () => {
        menu = renderer.create(
            <MemoryRouter>
                <UserDropdown />
            </MemoryRouter>
        )
    })
    expect(menu.root.findAllByType('li').map((li) => li.children[0])).toEqual([
        'Home',
        // 'Upload Panel',
        // 'My Uploads',
        // 'Admin Panel',
        'Log Out',
    ])
})

test('Test different paths', async () => {
    let menu;
    await renderer.act(async () => {
        menu = renderer.create(
            <MemoryRouter>
                <UserDropdown currentPath="upload" />
            </MemoryRouter>
        )
    })
    expect(menu.root.findAllByType('li').map(li => li.children[0])).not.toContain('Upload Panel')

    await renderer.act(async () => {
        menu.update(
            <MemoryRouter>
                <UserDropdown currentPath="recordingsManager" />
            </MemoryRouter>
        )
    })
    expect(menu.root.findAllByType('li').map(li => li.children[0])).not.toContain('My Uploads')
})

test('Test logged-out version', async () => {
    localStorage.setItem('loggedOut', true)
    const history = createMemoryHistory()
    await renderer.act(async () => {
        renderer.create(
            <Router history={history}>
                <UserDropdown />
            </Router>
        )
    })
    expect(history.location.pathname).toBe('/login')
})

test('Test navigation functionality', async () => {
    let menu
    const history = createMemoryHistory()
    history.push = jest.fn(() => {})
    await renderer.act(async () => {
        menu = renderer.create(
            <Router history={history}>
                <UserDropdown />
            </Router>
        )
    })
    const lis = menu.root.findAllByType('li')
    await renderer.act(async () => {
        for(let li of lis) {
            li.props.onClick()
        }
    })
    expect(history.push).toHaveBeenCalledTimes(lis.length)
    expect(history.push.mock.calls.map((call) => call[0])).toEqual(
        lis.map((li) => paths[li.children[0]])
    )
})