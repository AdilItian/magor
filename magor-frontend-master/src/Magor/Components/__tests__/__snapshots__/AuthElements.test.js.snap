// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic tests 1`] = `
<div
  className="dropDown"
  onMouseOut={[Function]}
  onMouseOver={[Function]}
  style={
    Object {
      "width": 220,
    }
  }
>
  <p>
    John Doe
  </p>
  <ul>
    <li
      onClick={[Function]}
    >
      Home
    </li>
    <li
      onClick={[Function]}
    >
      Upload Panel
    </li>
    <li
      onClick={[Function]}
    >
      My Uploads
    </li>
    <li
      onClick={[Function]}
    >
      Admin Panel
    </li>
    <li
      onClick={[Function]}
    >
      Log Out
    </li>
  </ul>
</div>
`;
